File: modules/implementation-python-runtime/src/main/resources/atomutil.py
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#  
#    http://www.apache.org/licenses/LICENSE-2.0
#    
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

# ATOM data conversion functions

from util import *
from elemutil import *
from xmlutil import *

# Convert a list of elements to a list of values representing an ATOM entry
def entryElementsToValues(e):
    lt = filter(selector((element, "'title")), e)
    t = "" if isNil(lt) else elementValue(car(lt))
    li = filter(selector((element, "'id")), e)
    i = "" if isNil(li) else elementValue(car(li))
    lc = filter(selector((element, "'content")), e)
    return (t, i, elementValue(car(lc)))

# Convert a list of elements to a list of values representing ATOM entries
def entriesElementsToValues(e):
    if isNil(e):
        return e
    return cons(entryElementsToValues(car(e)), entriesElementsToValues(cdr(e)))

# Convert a list of strings to a list of values representing an ATOM entry
def readATOMEntry(l):
    e = readXML(l)
    if isNil(e):
        return ()
    return entryElementsToValues(car(e))

# Convert a list of values representy an ATOM entry to a value
def entryValue(e):
    v = elementsToValues((caddr(e),))
    return cons(car(e), (cadr(e), cdr(car(v))))

# Return true if a list of strings represents an ATOM feed
def isATOMFeed(l):
    if isNil(l):
        return False
    if car(l)[0:5] != "<?xml":
        return False
    return contains(car(l), "<feed")

# Convert a list of strings to a list of values representing an ATOM feed
def readATOMFeed(l):
    f = readXML(l)
    if isNil(f):
        return ()
    t = filter(selector((element, "'title")), car(f))
    i = filter(selector((element, "'id")), car(f))
    e = filter(selector((element, "'entry")), car(f))
    if isNil(e):
        return (elementValue(car(t)), elementValue(car(i)))
    return cons(elementValue(car(t)), cons(elementValue(car(i)), entriesElementsToValues(e)))

# Convert an ATOM feed containing elements to an ATOM feed containing values
def feedValuesLoop(e):
    if (isNil(e)):
        return e
    return cons(entryValue(car(e)), feedValuesLoop(cdr(e)))

def feedValues(e):
    return cons(car(e), cons(cadr(e), feedValuesLoop(cddr(e))))

# Convert a list of values representy an ATOM entry to a list of elements
def entryElement(l):
    return (element, "'entry", (attribute, "'xmlns", "http://www.w3.org/2005/Atom"),
            (element, "'title", (attribute, "'type", "text"), car(l)),
            (element, "'id", cadr(l)),
            (element, "'content", (attribute, "'type", ("application/xml" if isList(caddr(l)) else "text")), caddr(l)),
            (element, "'link", (attribute, "'href", cadr(l))))

# Convert a list of values representing ATOM entries to a list of elements
def entriesElements(l):
    if isNil(l):
        return l
    return cons(entryElement(car(l)), entriesElements(cdr(l)))

# Convert a list of values representing an ATOM entry to an ATOM entry
def writeATOMEntry(l):
    return writeXML((entryElement(l),), True)

# Convert a list of values representing an ATOM feed to an ATOM feed
def writeATOMFeed(l):
    f = (element, "'feed", (attribute, "'xmlns", "http://www.w3.org/2005/Atom"),
            (element, "'title", (attribute, "'type", "text"), car(l)),
            (element, "'id", cadr(l)))
    if isNil(cddr(l)):
        return writeXML((f,), True)
    fe = append(f, entriesElements(cddr(l)))
    return writeXML((fe,), True)

# Convert an ATOM entry containing a value to an ATOM entry containing an item element
def entryValuesToElements(v):
    return cons(car(v), cons(cadr(v), valuesToElements((cons("'item", caddr(v)),))))

# Convert an ATOM feed containing values to an ATOM feed containing elements
def feedValuesToElementsLoop(v):
    if isNil(v):
        return v
    return cons(entryValuesToElements(car(v)), feedValuesToElementsLoop(cdr(v)))

def feedValuesToElements(v):
    return cons(car(v), cons(cadr(v), feedValuesToElementsLoop(cddr(v))))

File: modules/node-manager/src/test/resources/ui/atomutil.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 */

/**
 * ATOM data conversion functions.
 */
var atom = {};

/**
 * Convert a list of elements to a list of values representing an ATOM entry.
 */
atom.entryElementValues = function(e) {
    var lt = filter(selector(mklist(element, "'title")), e);
    var t = isNil(lt)? '' : elementValue(car(lt));
    var li = filter(selector(mklist(element, "'id")), e);
    var i = isNil(li)? '' : elementValue(car(li));
    var lc = filter(selector(mklist(element, "'content")), e);
    return append(mklist(element, "'entry", mklist(element, "'title", t), mklist(element, "'id", i)),
            isNil(lc)? mklist() : mklist(mklist(element, "'content", elementValue(car(lc)))))
};

/**
 * Convert a list of elements to a list of values representing ATOM entries.
 */
atom.entriesElementValues = function(e) {
    if (isNil(e))
        return e;
    return cons(atom.entryElementValues(car(e)), atom.entriesElementValues(cdr(e)));
};

/**
 * Return true if a list of strings represents an ATOM entry.
 */
atom.isATOMEntry = function(l) {
    if (!isXML(l))
        return false;
    return car(l).match('<entry') != null && car(l).match('<feed') == null && car(l).match('="http://www.w3.org/2005/Atom"') != null;
};

/**
 * Convert a DOM Document to a list of values representing an ATOM entry.
 */
atom.readATOMEntryDocument = function(doc) {
    var e = readXMLDocument(doc);
    if (isNil(e))
        return mklist();
    return mklist(atom.entryElementValues(car(e)));
};

/**
 * Convert a list of strings to a list of values representing an ATOM entry.
 */
atom.readATOMEntry = function(l) {
    return atom.readATOMEntryDocument(parseXML(l));
};

/**
 * Return true if a list of strings represents an ATOM feed.
 */
atom.isATOMFeed = function(l) {
    if (!isXML(l))
        return false;
    return car(l).match('<feed') != null && car(l).match('="http://www.w3.org/2005/Atom"') != null;
};

/**
 * Convert a DOM document to a list of values representing an ATOM feed.
 */
atom.readATOMFeedDocument = function(doc) {
    var f = readXMLDocument(doc);
    if (isNil(f))
        return mklist();
    var t = filter(selector(mklist(element, "'title")), car(f));
    var i = filter(selector(mklist(element, "'id")), car(f));
    var e = filter(selector(mklist(element, "'entry")), car(f));
    return mklist(append(
                mklist(element, "'feed", mklist(element, "'title", elementValue(car(t))), mklist(element, "'id", elementValue(car(i)))),
                atom.entriesElementValues(e)));
};

/**
 * Convert a list of strings to a list of values representing an ATOM feed.
 */
atom.readATOMFeed = function(l) {
    return atom.readATOMFeedDocument(parseXML(l));
};

/**
 * Convert a list of values representy an ATOM entry to a list of elements.
 */
atom.entryElement = function(l) {
    var title = elementValue(namedElementChild("'title", l));
    var id = elementValue(namedElementChild("'id", l));
    var content = namedElementChild("'content", l);
    var text = isNil(content)? false : elementHasValue(content);
    return append(append(
            mklist(element, "'entry", mklist(attribute, "'xmlns", "http://www.w3.org/2005/Atom"),
            mklist(element, "'title", mklist(attribute, "'type", "text"), title), mklist(element, "'id", id)),
            isNil(content)? mklist() :
                append(mklist(element, "'content", mklist(attribute, "'type", text? "text" : "application/xml")), text? mklist(elementValue(content)) : elementChildren(content))),
            mklist(element, "'link", mklist(attribute, "'href", id)));
};

/**
 * Convert a list of values representing ATOM entries to a list of elements.
 */
atom.entriesElements = function(l) {
    if (isNil(l))
        return l;
    return cons(atom.entryElement(car(l)), atom.entriesElements(cdr(l)));
};

/**
 * Convert a list of values representing an ATOM entry to an ATOM entry.
 */
atom.writeATOMEntry = function(ll) {
    var l = isNil(ll)? ll : car(ll);
    return writeXML(mklist(atom.entryElement(l)), true);
};

/**
 * Convert a list of values representing an ATOM feed to an ATOM feed.
 */
atom.writeATOMFeed = function(ll) {
    var l = isNil(ll)? ll : car(ll);
    var lt = filter(selector(mklist(element, "'title")), l);
    var t = isNil(lt)? '' : elementValue(car(lt));
    var li = filter(selector(mklist(element, "'id")), l);
    var i = isNil(li)? '' : elementValue(car(li));
    var f = mklist(element, "'feed", mklist(attribute, "'xmlns", "http://www.w3.org/2005/Atom"),
            mklist(element, "'title", mklist(attribute, "'type", "text"), car(l)),
            mklist(element, "'id", cadr(l)));

    // Write ATOM entries
    var le = filter(selector(mklist(element, "'entry")), l);
    if (isNil(le))
        return writeXML(mklist(f), true);

    // Write a single ATOM entry element with a list of values
    if (!isNil(le) && !isNil(car(le)) && isList(car(caddr(car(le))))) {
        var fe = append(f, atom.entriesElements(caddr(car(le))));
        return writeXML(mklist(fe), true);
    }

    // Write separate ATOM entry elements
    var fe = append(f, atom.entriesElements(le));
    return writeXML(mklist(fe), true);
};

